File: app/src/main/java/com/openvehicles/OVMS/api/ApiObservable.java
package com.openvehicles.OVMS.api;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import android.os.Handler;
import android.os.Message;

import com.openvehicles.OVMS.entities.CarData;

public class ApiObservable {
	private static ApiObservable sInstance;
	
	private final List<ApiObserver> mObservers = new ArrayList<ApiObserver>();
	private final NotifyOneHandler mHandler;
	private boolean mIsLoggedIn = false;
	
	public static ApiObservable get() {
		if (sInstance == null) {
			sInstance = new ApiObservable();
		}
		return sInstance; 
	}
	
	public ApiObservable() {
		mHandler = new NotifyOneHandler(this);		
	}

    public void addObserver(ApiObserver pObserver) {
        if (pObserver == null) {
            throw new NullPointerException();
        }
        synchronized (this) {
            if (!mObservers.contains(pObserver))
                mObservers.add(pObserver);
        }
    }

    public int countObservers() {
        return mObservers.size();
    }

    public synchronized void deleteObserver(ApiObserver pObserver) {
        mObservers.remove(pObserver);
    }

    public synchronized void deleteObservers() {
        mObservers.clear();
    }
    
    public void notifyOnBind(ApiService pService) {
        mIsLoggedIn = pService.isLoggedIn();
		int size = 0;
        ApiObserver[] arrays = null;
        synchronized (this) {
            size = mObservers.size();
            arrays = new ApiObserver[size];
            mObservers.toArray(arrays);
        }
		for (ApiObserver observer : arrays) {
			observer.onServiceAvailable(pService);
			observer.onServiceLoggedIn(pService, mIsLoggedIn);
		}
    }

	public void notifyLoggedIn(ApiService pService, boolean pIsLoggedIn) {
		// Only notify on state changes:
		if (mIsLoggedIn == pIsLoggedIn) return;
		mIsLoggedIn = pIsLoggedIn;

		// OK, notify observers:
		int size = 0;
		ApiObserver[] arrays = null;
		synchronized (this) {
			size = mObservers.size();
			arrays = new ApiObserver[size];
			mObservers.toArray(arrays);
		}
		for (ApiObserver observer : arrays) {
			observer.onServiceLoggedIn(pService, pIsLoggedIn);
		}
	}

    public void notifyUpdate(CarData pCarData) {
    	mHandler.notifyUpdate(pCarData);
    }
    
    private void notifyOneUpdate(CarData pCarData) {
        int size = 0;
        ApiObserver[] arrays = null;
        synchronized (this) {
            size = mObservers.size();
            arrays = new ApiObserver[size];
            mObservers.toArray(arrays);
        }
        if (arrays != null) {
            for (ApiObserver observer : arrays) {
                observer.update(pCarData);
            }
        }
    }
    
    private static class NotifyOneHandler extends Handler {
		private static final int WHAT = 100; 
    	private final WeakReference<ApiObservable> mParent;
		
		public NotifyOneHandler(ApiObservable pParent) {
			mParent = new WeakReference<ApiObservable>(pParent);
		}
		
		public void notifyUpdate(CarData pCarData) {
			Message msg = new Message();
			msg.what = WHAT;
			msg.obj = pCarData;
			
			removeMessages(WHAT, pCarData);
			sendMessageDelayed(msg, 300);
		}
		
		@Override
		public void handleMessage(Message msg) {
			mParent.get().notifyOneUpdate((CarData) msg.obj);
		}
	}
    
}

File: app/src/main/java/com/openvehicles/OVMS/api/ApiObservable.kt
package com.openvehicles.OVMS.api

import android.os.Handler
import android.os.Message
import com.openvehicles.OVMS.entities.CarData
import java.lang.ref.WeakReference

object ApiObservable {

    private val observers = mutableListOf<ApiObserver>()
    private val handler = NotifyOneHandler(this)
    private var isLoggedIn = false

    fun addObserver(observer: ApiObserver) {
        synchronized(this) {
            if (!observers.contains(observer)) observers.add(observer)
        }
    }

    fun countObservers(): Int {
        return observers.size
    }

    @Synchronized
    fun deleteObserver(observer: ApiObserver) {
        observers.remove(observer)
    }

    @Synchronized
    fun deleteObservers() {
        observers.clear()
    }

    fun notifyOnBind(service: ApiService) {
        isLoggedIn = service.isLoggedIn()
        var arrays: Array<ApiObserver>
        synchronized(this) {
            arrays = observers.toTypedArray()
        }
        for (observer in arrays) {
            observer.onServiceAvailable(service)
            observer.onServiceLoggedIn(service, isLoggedIn)
        }
    }

    fun notifyLoggedIn(service: ApiService?, isLoggedIn: Boolean) {
        // Only notify on state changes:
        if (this.isLoggedIn == isLoggedIn) {
            return
        }
        this.isLoggedIn = isLoggedIn

        // OK, notify observers:
        var arrays: Array<ApiObserver>
        synchronized(this) {
            arrays = observers.toTypedArray()
        }
        for (observer in arrays) {
            observer.onServiceLoggedIn(service, isLoggedIn)
        }
    }

    fun notifyUpdate(pCarData: CarData?) {
        handler.notifyUpdate(pCarData)
    }

    private fun notifyOneUpdate(carData: CarData) {
        var arrays: Array<ApiObserver>
        synchronized(this) {
            arrays = observers.toTypedArray()
        }
        for (observer in arrays) {
            observer.update(carData)
        }
    }


    private class NotifyOneHandler(parent: ApiObservable) : Handler() {

        private val parent = WeakReference(parent)

        fun notifyUpdate(pCarData: CarData?) {
            val msg = Message()
            msg.what = WHAT
            msg.obj = pCarData
            removeMessages(WHAT, pCarData)
            sendMessageDelayed(msg, 300)
        }

        override fun handleMessage(msg: Message) {
            parent.get()!!.notifyOneUpdate(msg.obj as CarData)
        }

        companion object {

            private const val WHAT = 100

        }
    }

}